app/questionnaire_state/answer.py

```diff
@@ -9,9 +9,33 @@ def __init__(self, id):
         # actual user input
         self.input = None
 
-    def update_state(self, user_input):
-        if self.id in user_input.keys():
-            self.input = user_input[self.id]
+    def update_state(self, user_input, schema_item):
+        # Do we have the item or it's containing block?
+        if schema_item.id != self.id:
+            schema_item = schema_item.questionnaire.get_item_by_id(self.id)
+
+        # Get the user input
+        self.input = schema_item.get_user_input(user_input)
+
```

Comment:
I was just suggesting extracting out the code below into a sub-method as it is specifically related to type checking, something like _validate_input() or _extract_type() would be viable method names. It just keeps the size of the method down, limits complexity and makes the code more readable.